From 728502bb31b7aa04caa03d1414adb8e6ed9bfe35 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Tue, 18 Feb 2025 20:00:51 +0800
Subject: [PATCH 070/123] mtk: mt76: mt7996: fix monitor mode cpu busy issue

Rx token id may be incorrect when scatter and gather is enabled.
We need to find the correct rxwi based on physical address.
However, it wastes lots of time if we go through every rxwi for
every packet, we only need to find rxwi when the physical address
in rxwi is different from the physical address in rx_dmad.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 dma.c  |  4 +++-
 mt76.h |  1 +
 tx.c   | 13 +++++++++++++
 3 files changed, 17 insertions(+), 1 deletion(-)

diff --git a/dma.c b/dma.c
index 1cac18d6..1ce69893 100644
--- a/dma.c
+++ b/dma.c
@@ -469,7 +469,9 @@ mt76_dma_get_buf(struct mt76_dev *dev, struct mt76_queue *q, int idx,
 		u32 token = FIELD_GET(MT_DMA_CTL_TOKEN, buf1);
 		struct mt76_txwi_cache *t;
 
-		if (*more) {
+		t = mt76_rx_token_find(dev, token);
+
+		if (*more && (!t || t->dma_addr != le32_to_cpu(desc->buf0))) {
 			spin_lock_bh(&dev->rx_token_lock);
 
 			idr_for_each_entry(&dev->rx_token, t, id) {
diff --git a/mt76.h b/mt76.h
index a175e033..9fc6baef 100644
--- a/mt76.h
+++ b/mt76.h
@@ -2122,6 +2122,7 @@ int mt76_token_consume(struct mt76_dev *dev, struct mt76_txwi_cache **ptxwi,
 		       u8 phy_idx);
 void __mt76_set_tx_blocked(struct mt76_dev *dev, bool blocked);
 struct mt76_txwi_cache *mt76_rx_token_release(struct mt76_dev *dev, int token);
+struct mt76_txwi_cache *mt76_rx_token_find(struct mt76_dev *dev, int token);
 int mt76_rx_token_consume(struct mt76_dev *dev, void *ptr,
 			  struct mt76_txwi_cache *r, dma_addr_t phys);
 int mt76_create_page_pool(struct mt76_dev *dev, struct mt76_queue *q);
diff --git a/tx.c b/tx.c
index edb5cef5..c0f4fca2 100644
--- a/tx.c
+++ b/tx.c
@@ -956,3 +956,16 @@ mt76_rx_token_release(struct mt76_dev *dev, int token)
 	return t;
 }
 EXPORT_SYMBOL_GPL(mt76_rx_token_release);
+
+struct mt76_txwi_cache *
+mt76_rx_token_find(struct mt76_dev *dev, int token)
+{
+	struct mt76_txwi_cache *t;
+
+	spin_lock_bh(&dev->rx_token_lock);
+	t = idr_find(&dev->rx_token, token);
+	spin_unlock_bh(&dev->rx_token_lock);
+
+	return t;
+}
+EXPORT_SYMBOL_GPL(mt76_rx_token_find);
-- 
2.45.2

